src/Microsoft.AspNetCore.Mvc.Razor/RazorPageActivator.cs

```diff
@@ -29,13 +31,35 @@ public class RazorPageActivator : IRazorPageActivator
         private readonly ConcurrentDictionary<Type, PageActivationInfo> _activationInfo;
         private readonly IModelMetadataProvider _metadataProvider;
 
+        // Common Singleton properties activated in a RazorPage.
+        private readonly IUrlHelperFactory _urlHelperFactory;
+        private Func<ViewContext, object> _urlHelperAccessor;
+
+        private readonly IJsonHelper _jsonHelper;
+        private Func<ViewContext, object> _jsonHelperAccessor;
+
+        private readonly DiagnosticSource _diagnosticSource;
+        private Func<ViewContext, object> _diagnosticSourceAccessor;
+
+        private readonly HtmlEncoder _htmlencoder;
+        private Func<ViewContext, object> _htmlencoderAccessor;
+
         /// <summary>
         /// Initializes a new instance of the <see cref="RazorPageActivator"/> class.
         /// </summary>
-        public RazorPageActivator(IModelMetadataProvider metadataProvider)
+        public RazorPageActivator(
+            IModelMetadataProvider metadataProvider,
+            IUrlHelperFactory urlHelperFactory,
+            IJsonHelper jsonHelper,
+            DiagnosticSource diagnosticSource,
+            HtmlEncoder htmlEncoder)
         {
             _activationInfo = new ConcurrentDictionary<Type, PageActivationInfo>();
             _metadataProvider = metadataProvider;
+            _urlHelperFactory = urlHelperFactory;
+            _jsonHelper = jsonHelper;
+            _diagnosticSource = diagnosticSource;
+            _htmlencoder = htmlEncoder;
```

Comment:
Suggest removing the lazy init of the accessors and doing those assignments here.